src/components/visiumQC/CDNAMeasurementQC.tsx

```diff
@@ -0,0 +1,303 @@
+ * @param validateMeasurementValue - Validate function for measurement value
+ * @constructor
+ */
+const CDNAMeasurementQC = ({
+  measurementName,
+  qcType,
+  onSave,
+  onError,
+  applySameValueForAlMeasurements,
+  stepIncrement,
+  initialMeasurementVal,
+  validateMeasurementValue,
+}: {
+  measurementName: string;
+  qcType: string;
+  onSave: () => void;
+  onError: (error: ClientError) => void;
+  applySameValueForAlMeasurements: boolean;
+  stepIncrement?: string;
+  initialMeasurementVal?: string;
+  validateMeasurementValue?: (value: string) => void;
+}) => {
+  const { values, errors, validateForm, setErrors } = useFormikContext<
+    VisiumQCData
+  >();
+
+  const { labwares, removeLabware } = useLabwareContext();
+
+  const stanCore = useContext(StanCoreContext);
+
+  const [slotMeasurements, setSlotMeasurements] = useState<
+    SlotMeasurementRequest[]
+  >([]);
+  const [validationError, setValidationError] = useState("");
+
+  /***State machine**/
+  const [current, send] = useMachine(
+    createFormMachine<
+      OpWithSlotMeasurementsRequest,
+      RecordOpWithSlotMeasurementsMutation
+    >().withConfig({
+      services: {
+        submitForm: (ctx, e) => {
+          if (e.type !== "SUBMIT_FORM") return Promise.reject();
+          return stanCore.RecordOpWithSlotMeasurements({
+            request: e.values,
+          });
+        },
+      },
+    })
+  );
+  const { serverError } = current.context;
+
+  /***
+   * Initializing SlotMeasurement values
+   * **/
+  const buildSlotMeasurements = React.useCallback(
+    (labware: LabwareFieldsFragment): SlotMeasurementRequest[] => {
+      return labware.slots.filter(isSlotFilled).map((slot) => ({
+        address: slot.address,
+        name: measurementName,
+        value: initialMeasurementVal ?? "",
+      }));
+    },
+    [initialMeasurementVal, measurementName]
+  );
+
+  /***
+   * When labwares changes, slotMeasurements need to be updated
+   */
+  useEffect(() => {
+    setSlotMeasurements(
+      labwares.length > 0 ? buildSlotMeasurements(labwares[0]) : []
+    );
+  }, [labwares, setSlotMeasurements, buildSlotMeasurements]);
+
+  /***
+   * Save(/Recording) operation completed.
+   * Notify the parent component with the outcome of Save operation
+   */
+  useEffect(() => {
+    if (current.matches("submitted")) {
+      onSave();
+    }
+    if (serverError) {
+      onError(serverError);
+    }
+  }, [current, serverError, onSave, onError]);
+
+  /***
+   * Reset form errors on QCType chnage
+   */
+  useEffect(() => {
+    if (!qcType) return;
+
+    setErrors({});
+  }, [qcType, setErrors]);
+
+  /***
+   * Callback to handle Save action
+   * **/
+  const handleSave = async () => {
+    //Validate form as the Save button is not implemented as a typical Formit Submit button
+    await validateForm();
+
+    //If errors are in form data, notify user
+    if (errors.slotMeasurements && errors.slotMeasurements.length > 0) {
```

Comment:
What if there are errors on other fields that aren't slotMeasurements?

Reply:
This is handled differently in new changes.